### PR TITLE
[SPARK-48152][BUILD] Make `spark-profiler` as a part of release and publish to maven central repo

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to
- make the module `spark-profiler` as a part of Spark release
- publish the module `spark-profiler` to `maven central repository`
- add instructions on how to compile supports `spark-profiler` in the doc `docs/building-spark.md` 

### Why are the changes needed?
1.The modules released in the current daily `spark-4.0.0` do not include `spark-profiler`. I believe that according to the current logic, the `spark-profiler` will not appear in the future official version of spark.

2.Align the compilation description of other modules in doc `docs/building-spark.md`, eg:
image

### Does this PR introduce _any_ user-facing change?
Yes, make it easy for users to use `spark-profiler` in the future version of Spark, instead of manually compiling `spark-profiler` based on source code.

### How was this patch tested?
- Pass GA.
- It is necessary to observe whether the daily snapshots `spark-profilter_2.13` generate
https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-profiler_2.13/4.0.0-SNAPSHOT/

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @panbingkun .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Ur, I agree with publishing this module.  However, I don't think we need to include this as Apache Spark Binary or `dev/deps`. This module is designed to be like `kafka` module, @panbingkun .

> [Only test] make spark-profiler publish snapshot [SPARK-48152][BUILD] Make the module spark-profiler as a part of Spark release

File: dev/test-dependencies.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ export LC_ALL=C
 # NOTE: These should match those in the release publishing script, and be kept in sync with
 #   dev/create-release/release-build.sh
 HADOOP_MODULE_PROFILES="-Phive-thriftserver -Pkubernetes -Pyarn -Phive \
-    -Pspark-ganglia-lgpl -Pkinesis-asl -Phadoop-cloud"
+    -Pspark-ganglia-lgpl -Pkinesis-asl -Phadoop-cloud -Pjvm-profiler"
```
Comment (dongjoon-hyun) [inline dev/test-dependencies.sh:34]: Like `Kafka` module, we should not include this here.

Comment (dongjoon-hyun) [reply dev/test-dependencies.sh:34]:
```
$ grep kafka dev/deps/spark-deps-hadoop-3-hive-2.3 | wc -l
       0
```

Comment (panbingkun) [reply dev/test-dependencies.sh:34]: > ~Like `Kafka` module, we should not include this here.~
> 
> Do you know how we skip Kafka module's dependency here?

Let me investigate it.

Comment (dongjoon-hyun) [reply dev/test-dependencies.sh:34]: Yes, it's identical with Apache Spark's Kafka module and Apache Spark Hadoop Cloud module

- https://mvnrepository.com/artifact/org.apache.spark/spark-sql-kafka-0-10_2.13/3.5.1
- https://mvnrepository.com/artifact/org.apache.spark/spark-hadoop-cloud_2.13/3.5.1

Comment (panbingkun) [reply dev/test-dependencies.sh:34]: Yeah, let me add detailed guide in some document.

File: connector/profiler/pom.xml [hunk 2]
```diff
@@ -44,7 +44,7 @@
     <dependency>
       <groupId>me.bechberger</groupId>
       <artifactId>ap-loader-all</artifactId>
-      <version>3.0-8</version>
+      <version>3.0-9</version>
```
Comment (dongjoon-hyun) [inline connector/profiler/pom.xml:47]: Could you proceed this dependency update separately?

File: dev/test-dependencies.sh (hunk 1 above)
Comment (panbingkun) [reply dev/test-dependencies.sh:34]: > ~Like `Kafka` module, we should not include this here.~
> 
> Do you know how we skip Kafka module's dependency here?

I already know about it. It is implemented through `maven scope` overloading. Let me try.

File: connector/profiler/pom.xml (hunk 2 above)
Comment (panbingkun) [reply connector/profiler/pom.xml:47]: Okay

File: docs/building-spark.md [hunk 3]
```diff
@@ -117,6 +117,13 @@ where `spark-streaming_{{site.SCALA_BINARY_VERSION}}` is the `artifactId` as def
 
     ./build/mvn -Pconnect -DskipTests clean package
 
+## Building with JVM Profile support
+
+    ./build/mvn -Pjvm-profiler -DskipTests clean package
+
+**Note:** The `jvm-profiler` profile builds the assembly without including the dependency `ap-loader`,
+you can download it manually from maven central repo and use it together with `spark-profiler_{{site.SCALA_BINARY_VERSION}}`.
+
```
Comment (panbingkun) [inline docs/building-spark.md:126]: The actual effect is like this:
image

File: connector/profiler/pom.xml (hunk 2 above)
Comment (panbingkun) [reply connector/profiler/pom.xml]: A separate PR:
https://github.com/apache/spark/pull/46427

File: connector/profiler/pom.xml [hunk 4]
```diff
@@ -45,6 +48,7 @@
       <groupId>me.bechberger</groupId>
       <artifactId>ap-loader-all</artifactId>
       <version>3.0-8</version>
+      <scope>provided</scope>
```
Comment (dongjoon-hyun) [inline connector/profiler/pom.xml:51]: cc @parthchandra , too

File: connector/profiler/README.md [hunk 5]
```diff
@@ -23,7 +23,7 @@ Code profiling is currently only supported for
 To get maximum profiling information set the following jvm options for the executor :
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply connector/profiler/README.md:23]: cc @parthchandra

File: connector/profiler/pom.xml (hunk 4 above)
Comment (parthchandra) [reply connector/profiler/pom.xml:51]: It's great to include this feature in the spark release. I feel though, that if we are making it available in the release the ap-loader dependency should be included as well. (Currently, if we build with the jvm-profiler profile, the dependency is included in the build). 
Surprisingly, the jar file is not very large (I have version 2.9.7 and it is only 830K).
Either way, including this in the release is already a big step so I'm fine with asking users to download the jar.

**Comment by dongjoon-hyun** [comment]
Could you revise the PR title, @panbingkun ?

**Comment by panbingkun** [comment]
> I merged the following. Could you rebase this PR?
> 
> * [[SPARK-48165][BUILD] Update `ap-loader` to 3.0-9 #46427](https://github.com/apache/spark/pull/46427)

Done.

**Comment by panbingkun** [comment]
> Could you revise the PR title, @panbingkun ?

Done.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM for Apache Spark 4.0.0-preview.

- Thank you for completing area, @panbingkun .
- Thank you for review and understanding the current usage, @parthchandra .

Let's bring this into 4.0.0-preview and get more feedback. We can revise later before 4.0.0.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by panbingkun** [comment]
> Let's bring this into 4.0.0-preview and get more feedback. We can revise later before 4.0.0.

Thanks. ❤️

**Comment by panbingkun** [comment]
It's so happy to observe that `maven's snapshot repo` has `spark-profiler`:
https://repository.apache.org/content/repositories/snapshots/org/apache/spark/spark-profiler_2.13/4.0.0-SNAPSHOT/
image

**Pull request closed without merging**